### PR TITLE
respect max execution time limit for transitions plugin

### LINKED ISSUE #15609: Limit max execution time of transitions feature
We should apply `live_query_max_execution_time` setting also for transition queries since it is basically also a live query. Ideally, if the limit is reached, similar to visitor log we would then show a message if the limit is reached and suggest to shorten the date range for example, or if a segment is selected mention could be caused by this.

refs https://github.com/matomo-org/matomo/issues/14762

### PR DESCRIPTION
**Pull request by sgiehl**
fixes #15609

### CODE REVIEW

File: core/DbHelper.php [hunk 1]
```diff
@@ -212,6 +212,33 @@ public static function deleteArchiveTables()
...
+
+            $timeInMs = $limit * 1000;
+            $timeInMs = (int) $timeInMs;
+            $maxExecutionTimeHint = ' /*+ MAX_EXECUTION_TIME('.$timeInMs.') */ ';
+
+            $sql = substr_replace($sql, 'SELECT ' . $maxExecutionTimeHint, $pos, strlen('SELECT'));
+        }
+
+        return $sql;
+    }
```
Comment (diosmosis) [inline core/DbHelper.php]: Can we add some unit tests for this logic?

File: plugins/Live/Model.php [hunk 2]
```diff
@@ -130,57 +131,58 @@ private function executeLogVisitsQuery($sql, $bind, $segment, $dateStart, $dateE
...
 	 * @param $dateStart
 	 * @param $dateEnd
 	 * @param $minTimestamp
 	 * @param $limit
+	 * @param $parameters
 	 *
 	 * @throws MaxExecutionTimeExceededException
 	 */
-    public function handleMaxExecutionTimeError($readerDb, $e, $sql, $bind, $segment, $dateStart, $dateEnd, $minTimestamp, $limit)
+    public static function handleMaxExecutionTimeError($readerDb, $e, $segment, $dateStart, $dateEnd, $minTimestamp, $limit, $parameters)
```
Comment (diosmosis) [inline plugins/Live/Model.php]: If unit tests don't exist for this method can we add some? I'm not sure if it is easy to do however. Some sort of test covering this behavior would be good IMO

**Comment by diosmosis** [comment]
Code looks good to me but I'm not too familiar with the related code, so I think @tsteur may want to take a look as well

File: plugins/Transitions/API.php [hunk 3]
```diff
@@ -81,25 +84,39 @@ public function getTransitionsForAction($actionName, $actionType, $idSite, $peri
...
-                - $this->getTotalTransitionsToFollowingActions()
-                - $report['pageMetrics']['loops'];
+            // derive the number of exits from the other metrics
+            if ($parts == 'all') {
+                $report['pageMetrics']['exits'] = $report['pageMetrics']['pageviews']
+                    - $this->getTotalTransitionsToFollowingActions()
+                    - $report['pageMetrics']['loops'];
+            }
+        } catch (\Exception $e) {
+            Model::handleMaxExecutionTimeError(
```
Comment (tsteur) [inline plugins/Transitions/API.php]: Is there a chance that if the query doesn't fail because of max execution error, the exception is not thrown?

**Comment by tsteur** [comment]
Looks good. Left one comment. Haven't tested it though. Be good to make sure some error message will be shown in the UI if the query fails because of a timeout similar to Visitor Log. I suppose that's what happens automatically anyway. Just need to see how it behaves when it's used in transitions eg through a row action

File: plugins/Live/Model.php (hunk 2 above)
Comment (sgiehl) [reply plugins/Live/Model.php]: that method itself is tested here
https://github.com/matomo-org/matomo/blob/148e049bc7665af470514a8f0a573973fc59a52b/plugins/Live/tests/Integration/ModelTest.php#L74-L126

File: plugins/Transitions/API.php (hunk 3 above)
Comment (sgiehl) [reply plugins/Transitions/API.php]: oh sure. there should be a `throw $e` after that. Will add...

**Comment by sgiehl** [comment]
@tsteur it show the exception in the overlay if loading a transition through a row action:

image

File: core/DbHelper.php (hunk 1 above)
Comment (sgiehl) [reply core/DbHelper.php]: added some simple tests

File: plugins/Live/Model.php (hunk 2 above)
Comment (tsteur) [reply plugins/Live/Model.php]: Wouldn't this test need to be adjusted? Not seeing it in this PR? Would expect seeing this new parameter and calling it static @sgiehl ?

**Comment by tsteur** [comment]
@sgiehl I just tested it and it doesn't work. That's because the ranking query will be applied and therefore it's not after the first `select`.
Not sure how it works for you? Cause they all have ranking query set?

This is my setting
> live_query_max_execution_time = 0.001

**Comment by tsteur** [comment]
btw we could in worst case move this out of 3.13.4 into 3.13.5 if we wanted to release 3.13.4 earlier

**Comment by sgiehl** [comment]
@tsteur updated the code so it works with the ranking query correctly.

**Pull request merged**